x-pack/plugins/reporting/server/plugin.ts

```diff
@@ -22,61 +20,83 @@ export class ReportingPlugin
   private readonly initializerContext: PluginInitializerContext<ReportingConfigType>;
   private logger: LevelLogger;
   private reportingCore?: ReportingCore;
-  private config$: Observable<ReportingConfigType>;
+
+  // Setup some observables for modules that need to await setup/start
+  public readonly setup$ = new Rx.Subject<boolean>();
+  public readonly start$ = new Rx.Subject<boolean>();
 
   constructor(context: PluginInitializerContext<ReportingConfigType>) {
     this.logger = new LevelLogger(context.logger.get());
     this.initializerContext = context;
-    this.config$ = context.config.create<ReportingConfigType>();
   }
 
-  public async setup(core: CoreSetup, plugins: ReportingSetupDeps) {
+  public setup(core: CoreSetup, plugins: ReportingSetupDeps) {
     const { elasticsearch, http } = core;
     const { licensing, security } = plugins;
     const { initializerContext: initContext } = this;
     const router = http.createRouter();
     const basePath = http.basePath.get;
 
-    const coreConfig = await createConfig$(core, this.config$, this.logger)
-      .pipe(first())
-      .toPromise(); // apply computed defaults to config
-    const reportingConfig = buildConfig(initContext, core, coreConfig); // combine kbnServer configs
-    this.reportingCore = new ReportingCore(reportingConfig);
-
-    const browserDriverFactory = await createBrowserDriverFactory(reportingConfig, this.logger);
-
-    this.reportingCore.pluginSetup({
-      browserDriverFactory,
-      elasticsearch,
-      licensing,
-      basePath,
-      router,
-      security,
-    });
+    // async background setup
+    (async () => {
+      const config = await buildConfig(initContext, core, this.logger);
+      const reportingCore = new ReportingCore(config);
+
+      reportingCore.pluginSetup({
+        elasticsearch,
+        licensing,
+        basePath,
+        router,
+        security,
+      });
 
-    runValidations(reportingConfig, elasticsearch, browserDriverFactory, this.logger);
-    registerReportingUsageCollector(this.reportingCore, plugins);
-    registerRoutes(this.reportingCore, this.logger);
+      registerReportingUsageCollector(reportingCore, plugins);
+      registerRoutes(reportingCore, this.logger);
```

Comment:
`setup` shouldn't be called after setup lifecycle completed. the platform is going to start throwing the exception. you can pass custom arguments via request handler context instead https://github.com/elastic/kibana/blob/a85bcc6d4efaada428240833c2cba942ec1acb9e/src/core/MIGRATION.md#declare-a-custom-scoped-service

Another downside of the current solution, that reporting endpoints will return 404, instead of a more meaningful message that the service is not ready yet.

Reply:
Filed: https://github.com/elastic/kibana/issues/68927 which includes an example of using route handler context